pkg/federate/kubefed/cluster.go

```diff
@@ -1,9 +1,299 @@
+	}
+
+	kubeConfig, err := clientcmd.BuildConfigFromFlags(kubeFedCluster.Spec.APIEndpoint, "")
+	if err != nil {
+		return nil, err
+	}
+
+	kubeConfig.CAData = kubeFedCluster.Spec.CABundle
+	kubeConfig.BearerToken = string(token)
+
+	return kubeConfig, nil
+}
+
+func newKubeFedClusterInformer(kubeFedConfig *rest.Config) (cache.SharedIndexInformer, error) {
+	var obj runtime.Object = &fedv1.KubeFedCluster{}
+
+	gvk, err := apiutil.GVKForObject(obj, scheme.Scheme)
+	if err != nil {
+		return nil, fmt.Errorf("error finding GroupVersionKind for %T: %v", obj, err)
+	}
+
+	klog.V(3).Infof("GroupVersionKind for %T: %#v", obj, gvk)
+
+	mapper, err := apiutil.NewDiscoveryRESTMapper(kubeFedConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error creating RESTMapper from config: %v", err)
+	}
+
+	mapping, err := mapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+	if err != nil {
+		return nil, fmt.Errorf("error creating RESTMapping for %#v: %v", gvk, err)
+	}
+
+	client, err := apiutil.RESTClientForGVK(gvk, kubeFedConfig, scheme.Codecs)
+	if err != nil {
+		return nil, fmt.Errorf("error creating RESTClient for %#v: %v", gvk, err)
+	}
+
+	klog.V(3).Infof("Got RESTClient: %#v", client)
+
+	return cache.NewSharedIndexInformer(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				var timeout time.Duration
+				if options.TimeoutSeconds != nil {
+					timeout = time.Duration(*options.TimeoutSeconds) * time.Second
+				}
+
+				result := &fedv1.KubeFedClusterList{}
+				err := client.Get().Namespace(kubeFedNamespace).Resource(mapping.Resource.Resource).
+					VersionedParams(&options, scheme.ParameterCodec).Timeout(timeout).Do().Into(result)
+				return result, err
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				var timeout time.Duration
+				if options.TimeoutSeconds != nil {
+					timeout = time.Duration(*options.TimeoutSeconds) * time.Second
+				}
+
+				options.Watch = true
+				return client.Get().Namespace(kubeFedNamespace).Resource(mapping.Resource.Resource).
+					VersionedParams(&options, scheme.ParameterCodec).Timeout(timeout).Watch()
+			},
+		},
+		obj,
+		0, // Providing 0 duration to an informer indicates that resync should be delayed as long as possible
+		cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
+	), nil
+}
+
+func (c *clusterWatcher) onAdd(clusterID string, kubeConfig *rest.Config) {
+	c.enqueueEvent(func() {
+		c.handler.OnAdd(clusterID, kubeConfig)
+	}, clusterID, "Add")
+}
+
+func (c *clusterWatcher) onUpdate(clusterID string, kubeConfig *rest.Config) {
+	c.enqueueEvent(func() {
+		c.handler.OnUpdate(clusterID, kubeConfig)
+	}, clusterID, "Update")
+}
+
+func (c *clusterWatcher) onRemove(clusterID string) {
+	c.enqueueEvent(func() {
+		c.handler.OnRemove(clusterID)
+	}, clusterID, "Remove")
+}
+
+func (c *clusterWatcher) enqueueEvent(event func(), clusterID string, eventType string) {
```

Comment:
I did that to try to avoid one watcher from blocking event delivery to other watchers. Also it allows to identify a slow or blocked watcher inn the logs as you mentioned.